TokenBucket.php

```diff
@@ -0,0 +1,135 @@
+<?
+
+namespace iFixit\TokenBucket;
+
+use \DateTime;
+use \DateInterval;
+use \InvalidArgumentException;
+
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'TokenRate.php';
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'Backend.php';
+
+/**
+ * Implements the token bucket algorithm and stores tokens in storage.
+ */
+class TokenBucket {
+   // String to identify this bucket uniquely in storage.
+   private $identifier;
+   // TokenRate is when your parents try to ration your arcade spending.
+   private $rate;
+   // Where the token is to be stored.
+   private $storage;
+
+   /**
+    * @param $identifier string to uniqely store the bucket.
+    */
+   public function __construct($key, Backend $backend, TokenRate $rate) {
+      if (!is_string($key)) {
+         throw new InvalidArgumentException("identifier must be a string");
+      }
+
+      $this->key = $key;
+      $this->backend = $backend;
+      $this->rate = $rate;
+   }
+
+   /**
+    * Tries to consume a token, if a token can't be consumed, then a DateTime
+    * is returned for when the next token will be available.
+    *
+    * @return array index 0 being a boolean whether or not a token was consumed
+    *               index 1 being a timestamp of when there will be
+    *               enough tokens to consume the amount requested.
+    */
+   public function consume($amount) {
+      if (!is_int($amount)) {
+         throw new InvalidArgumentException("amount must be an int");
+      }
+
+      list($tokens, $lastConsume) = $this->getTokenCountHelper();
+      $tokens -= $amount;
+      if ($tokens < 0) {
+         return [false, $this->readyTime($amount, $tokens, $lastConsume)];
+      }
+
+      $now = new DateTime();
+      $this->storeTokens($tokens, $now->getTimestamp());
+      return [true, $now->getTimestamp()];
+   }
+
+   public function getTokenCount() {
+      list($tokens, $_) = $this->getTokenCountHelper();
+      return $tokens;
+   }
+
+   /**
+    * @return array index 0 the number of tokens in the bucket
+    *               index 1 the last consumtion of a token, this will be the
+    *               current time if there wasn't a stored bucket.
+    */
+   private function getTokenCountHelper() {
```

Comment:
Sure, mostly it's just not a function that should be in the public API, but is still needed and still gets tokens. So it is "helping" get tokens.